### PR TITLE
Store color tokens as SwiftUI `Color` literals, instead of `UIColor`

### PR DESCRIPTION
**Pull request by mischreiber**
### Platforms Impacted
- [x] iOS
- [x] visionOS
- [ ] macOS

### Description of changes

Back in #1655, we moved internal storage of various token color types, including `GlobalTokens`, to use `UIColor` rather than our custom `FluentUI.ColorValue` and `FluentUI.DynamicColor` type. The goal of this was to ease consumption of our toolkit for our (then) mostly-UIKit-based customers. However, in the span of a single short year, we've seen SwiftUI usage and requests at an unprecedented level, and unfortunately our UIKit-based types have been causing a variety of minor issues (not the least of which being an inability to share much code with macOS).

To modernize and prepare for a more SwiftUI-based future, this PR rethinks that approach and aims for a better middle-of-the-road solution.

* `GlobalTokens` will now be stored and provided as simple `SwiftUI.Color` types. These can be used in SwiftUI directly. We will provide automatic UIKit wrappers in a new `GlobalTokens+UIKit` extension.
* `FluentTheme+Tokens` will now fully separate color *storage* from color *availability*:
  * Raw value storage will return to `DynamicColor`, which has been converted to a fully internal type.
  * Public exposure of these alias tokens will again be available as either a `SwiftUI.Color` wrapped type, or a backwards-compatible `UIColor` API set in `FluentTheme+UIKit.swift`.

In addition, this change finally removes the deprecated `AliasTokens` class and its public accessors in `FluentTheme`. These APIs have been deprecated since October 2023 (see #1914), and any client still using them should have already transitioned to the new equivalent APIs.

### Binary change

Total increase: 171,592 bytes
Total decrease: -820,904 bytes
| File | Before | After | Delta |
|------|-------:|------:|------:|
| Total | 31,638,440 bytes | 30,989,128 bytes | 🎉 -649,312 bytes |
<details>
<summary> Full breakdown </summary>

| File | Before | After | Delta |
|------|-------:|------:|------:|
| FluentTheme+Tokens.o | 143,184 bytes | 190,032 bytes | ⚠️ 46,848 bytes |
| Color+DynamicColor.o | 0 bytes | 36,392 bytes | ⚠️ 36,392 bytes |
| FluentTheme+UIKit.o | 0 bytes | 27,024 bytes | ⚠️ 27,024 bytes |
| GlobalTokens+UIKit.o | 0 bytes | 19,392 bytes | ⚠️ 19,392 bytes |
| ControlTokenSet.o | 67,808 bytes | 76,536 bytes | ⚠️ 8,728 bytes |
| FluentTheme.o | 206,016 bytes | 212,248 bytes | ⚠️ 6,232 bytes |
| BadgeLabelTokenSet.o | 49,264 bytes | 53,832 bytes | ⚠️ 4,568 bytes |
| TableViewCellTokenSet.o | 126,512 bytes | 130,176 bytes | ⚠️ 3,664 bytes |
| CommandBarCommandGroupsView.o | 202,304 bytes | 205,912 bytes | ⚠️ 3,608 bytes |
| ActivityIndicatorTokenSet.o | 60,080 bytes | 63,632 bytes | ⚠️ 3,552 bytes |
| HeadsUpDisplayTokenSet.o | 53,664 bytes | 56,912 bytes | ⚠️ 3,248 bytes |
| CommandBarTokenSet.o | 67,800 bytes | 71,048 bytes | ⚠️ 3,248 bytes |
| ShimmerLinesView.o | 107,048 bytes | 109,792 bytes | ⚠️ 2,744 bytes |
| PageCardPresenterController.o | 147,864 bytes | 149,232 bytes | ⚠️ 1,368 bytes |
| FluentUIFramework.o | 77,952 bytes | 78,584 bytes | ⚠️ 632 bytes |
| DateTimePickerView.o | 211,256 bytes | 211,312 bytes | ⚠️ 56 bytes |
| BottomSheetTokenSet.o | 48,192 bytes | 48,248 bytes | ⚠️ 56 bytes |
| CalendarViewWeekdayHeadingView.o | 77,880 bytes | 77,928 bytes | ⚠️ 48 bytes |
| SearchBarTokenSet.o | 82,856 bytes | 82,904 bytes | ⚠️ 48 bytes |
| PillButtonTokenSet.o | 88,024 bytes | 88,072 bytes | ⚠️ 48 bytes |
| CalendarViewDayTodayCell.o | 37,608 bytes | 37,640 bytes | ⚠️ 32 bytes |
| DateTimePickerController.o | 143,120 bytes | 143,144 bytes | ⚠️ 24 bytes |
| DatePickerController.o | 327,512 bytes | 327,536 bytes | ⚠️ 24 bytes |
| DateTimePickerViewComponent.o | 139,104 bytes | 139,112 bytes | ⚠️ 8 bytes |
| TokenizedControlView.o | 48,160 bytes | 48,152 bytes | 🎉 -8 bytes |
| CardView.o | 243,000 bytes | 242,232 bytes | 🎉 -768 bytes |
| FluentButtonToggleStyle.o | 66,808 bytes | 65,648 bytes | 🎉 -1,160 bytes |
| PopupMenuItem.o | 129,528 bytes | 127,544 bytes | 🎉 -1,984 bytes |
| CommandBarButton.o | 94,640 bytes | 92,384 bytes | 🎉 -2,256 bytes |
| ActionsCell.o | 163,032 bytes | 160,720 bytes | 🎉 -2,312 bytes |
| NavigationBarTokenSet.o | 74,432 bytes | 71,328 bytes | 🎉 -3,104 bytes |
| FocusRingView.o | 843,344 bytes | 839,960 bytes | 🎉 -3,384 bytes |
| AvatarTitleView.o | 187,168 bytes | 183,776 bytes | 🎉 -3,392 bytes |
| SideTabBar.o | 210,360 bytes | 206,952 bytes | 🎉 -3,408 bytes |
| PopupMenuItemCell.o | 165,800 bytes | 162,376 bytes | 🎉 -3,424 bytes |
| ShimmerView.o | 202,360 bytes | 198,912 bytes | 🎉 -3,448 bytes |
| PeoplePicker.o | 299,568 bytes | 296,080 bytes | 🎉 -3,488 bytes |
| BooleanCell.o | 87,128 bytes | 83,640 bytes | 🎉 -3,488 bytes |
| ActivityIndicator.o | 162,104 bytes | 158,608 bytes | 🎉 -3,496 bytes |
| UINavigationItem+Navigation.o | 80,688 bytes | 77,088 bytes | 🎉 -3,600 bytes |
| PersonaButtonCarousel.o | 171,608 bytes | 168,008 bytes | 🎉 -3,600 bytes |
| SegmentPillButton.o | 92,232 bytes | 88,616 bytes | 🎉 -3,616 bytes |
| BadgeView.o | 633,832 bytes | 629,968 bytes | 🎉 -3,864 bytes |
| CenteredLabelCell.o | 87,168 bytes | 83,160 bytes | 🎉 -4,008 bytes |
| TableViewHeaderFooterView.o | 283,440 bytes | 279,368 bytes | 🎉 -4,072 bytes |
| Separator.o | 74,248 bytes | 70,128 bytes | 🎉 -4,120 bytes |
| BadgeLabel.o | 64,072 bytes | 59,824 bytes | 🎉 -4,248 bytes |
| AvatarGroup.o | 421,760 bytes | 417,456 bytes | 🎉 -4,304 bytes |
| DrawerController.o | 488,504 bytes | 484,192 bytes | 🎉 -4,312 bytes |
| TwoLineTitleView.o | 260,800 bytes | 256,368 bytes | 🎉 -4,432 bytes |
| IndeterminateProgressBar.o | 234,784 bytes | 230,320 bytes | 🎉 -4,464 bytes |
| PersonaButton.o | 257,512 bytes | 252,952 bytes | 🎉 -4,560 bytes |
| DrawerShadowView.o | 78,392 bytes | 73,816 bytes | 🎉 -4,576 bytes |
| BottomCommandingController.o | 844,512 bytes | 839,808 bytes | 🎉 -4,704 bytes |
| ResizingHandleView.o | 74,008 bytes | 69,120 bytes | 🎉 -4,888 bytes |
| BadgeField.o | 580,568 bytes | 575,552 bytes | 🎉 -5,016 bytes |
| ListItem.o | 269,208 bytes | 263,888 bytes | 🎉 -5,320 bytes |
| NavigationBar.o | 559,712 bytes | 554,208 bytes | 🎉 -5,504 bytes |
| HeadsUpDisplay.o | 258,952 bytes | 253,224 bytes | 🎉 -5,728 bytes |
| BottomSheetController.o | 531,728 bytes | 525,872 bytes | 🎉 -5,856 bytes |
| TabBarItemView.o | 225,320 bytes | 219,440 bytes | 🎉 -5,880 bytes |
| SegmentedControl.o | 336,296 bytes | 330,216 bytes | 🎉 -6,080 bytes |
| PillButton.o | 170,176 bytes | 162,184 bytes | 🎉 -7,992 bytes |
| FluentTextField.o | 200,968 bytes | 192,080 bytes | 🎉 -8,888 bytes |
| Avatar.o | 614,232 bytes | 604,704 bytes | 🎉 -9,528 bytes |
| SearchBar.o | 378,584 bytes | 368,952 bytes | 🎉 -9,632 bytes |
| TooltipView.o | 214,008 bytes | 204,096 bytes | 🎉 -9,912 bytes |
| FluentNotification.o | 695,360 bytes | 684,792 bytes | 🎉 -10,568 bytes |
| FluentButtonStyle.o | 157,456 bytes | 146,280 bytes | 🎉 -11,176 bytes |
| CardNudge.o | 448,472 bytes | 437,168 bytes | 🎉 -11,304 bytes |
| TableViewCell.o | 838,896 bytes | 825,832 bytes | 🎉 -13,064 bytes |
| Label.o | 131,744 bytes | 118,504 bytes | 🎉 -13,240 bytes |
| Button.o | 213,136 bytes | 198,504 bytes | 🎉 -14,632 bytes |
| UIColor+Extensions.o | 81,624 bytes | 61,832 bytes | 🎉 -19,792 bytes |
| DynamicColor.o | 49,720 bytes | 0 bytes | 🎉 -49,720 bytes |
| __.SYMDEF | 4,908,536 bytes | 4,850,192 bytes | 🎉 -58,344 bytes |
| AliasTokens.o | 89,720 bytes | 0 bytes | 🎉 -89,720 bytes |
| GlobalTokens.o | 537,072 bytes | 199,552 bytes | 🎉 -337,520 bytes |
</details>

### Verification

* Ensured a variety of UIKit and SwiftUI components look and behave the same as before, in both light and dark modes.
* Added a new SwiftUI demo for alias tokens that bypasses UIKit entirely, to ensure that colors render properly in our new `Color+DynamicColor` extension.
* Added new unit tests to ensure our color conversions behave as expected.

<details>
<summary>Visual Verification</summary>

| UIKit                                       | SwiftUI                                      |
|----------------------------------------------|--------------------------------------------|
| UIKit | SwiftUI |

Simulator Screen Recording - iPhone 15 Pro - 2024-05-02 at 14 20 00

DarkElevated

</details>

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [x] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [x] Objective-C exposure (provide it only if needed)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/fluentui-apple/pull/2014)

### CODE REVIEW

File: ios/FluentUI/Extensions/Color+DynamicColor.swift [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    /// - Parameter light: The default `Color` for a light context. Required.
+    /// - Parameter dark: The override `Color` for a dark context. Optional.
+    /// - Parameter darkElevated: The override `Color` for a dark elevated context. Optional.
+    init(light: Color,
+         dark: Color? = nil,
+         darkElevated: Color? = nil) {
+
+        let dynamicColor = DynamicColor(light: light, dark: dark, darkElevated: darkElevated)
+        if #available(iOS 17, *) {
+            self.init(dynamicColor)
```
Comment (mischreiber) [inline ios/FluentUI/Extensions/Color+DynamicColor.swift:37]: This great new API was added in iOS 17, and allows you to create a `Color` with custom resolution logic. See the iOS 17-only extension below on `DynamicColor` that adds the new func `resolve(in environment: EnvironmentValues) -> Color.Resolved`.

File: ios/xcode/FluentUITests.xcconfig [hunk 2]
```diff
@@ -8,6 +8,7 @@ CODE_SIGN_STYLE = Automatic
 DEVELOPMENT_TEAM = UBF8T346G9
 INFOPLIST_FILE = FluentUI.Tests/Info.plist
 LD_RUNPATH_SEARCH_PATHS = $(inherited) @executable_path/Frameworks @loader_path/Frameworks
+OTHER_LDFLAGS = -ObjC
```
Comment (mischreiber) [inline ios/xcode/FluentUITests.xcconfig:11]: Without this flag, our unit tests won't pick up the methods in `FluentTheme+UIKit.swift`. We already use it for building our static lib and application, but were missing it here in the unit tests.

**Review by laminesm — COMMENTED** [review]
So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.

**Comment by mischreiber** [comment]
> So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.

That's correct. The Objective-C `-[MSFFluentTheme colorForToken:]` continues to be `UIColor` only. We could use different names for each, but I'm inclined towards leaving the existing name for `UIColor`, to minimize API breakage, and I couldn't think of a good name for the SwiftUI `Color` version. I'm open to suggestions though!

I chose to make `DynamicColor` internal to keep the usage of our API simple: all our public APIs only deal with existing, understood color objects and structs. One less thing to have to learn to use our stuff!

**Review by laminesm — APPROVED** [review]

**Comment by joannaquu** [comment]
> > So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.
> 
> That's correct. The Objective-C `-[MSFFluentTheme colorForToken:]` continues to be `UIColor` only. We could use different names for each, but I'm inclined towards leaving the existing name for `UIColor`, to minimize API breakage, and I couldn't think of a good name for the SwiftUI `Color` version. I'm open to suggestions though!
> 
> I chose to make `DynamicColor` internal to keep the usage of our API simple: all our public APIs only deal with existing, understood color objects and structs. One less thing to have to learn to use our stuff!

this might be a dumb question but why would we need `Color` in objc? wouldn't we only need that in swiftui?

**Review by joannaquu — APPROVED** [review]

File: ios/FluentUI/Core/Theme/FluentTheme+visionOS.swift [hunk 3]
```diff
@@ -3,57 +3,58 @@
...
-        let visionColor: UIColor
+    static func defaultColor_visionOS(_ token: FluentTheme.ColorToken) -> DynamicColor {
+        let visionColorDark: Color
 
         // Apply overrides as needed. Note that visionOS only supports one mode, so there's no
         // need to provide multiple values (e.g. light + dark, elevated, etc).
         switch token {
         case .foreground1:
-            visionColor = .white
+            visionColorDark = GlobalTokens.neutralColor(.white)
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/FluentTheme+visionOS.swift:17]: Ah that was part of an experiment. I'll consolidate them!

**Comment by mischreiber** [comment]
> > > So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.
> > 
> > 
> > That's correct. The Objective-C `-[MSFFluentTheme colorForToken:]` continues to be `UIColor` only. We could use different names for each, but I'm inclined towards leaving the existing name for `UIColor`, to minimize API breakage, and I couldn't think of a good name for the SwiftUI `Color` version. I'm open to suggestions though!
> > I chose to make `DynamicColor` internal to keep the usage of our API simple: all our public APIs only deal with existing, understood color objects and structs. One less thing to have to learn to use our stuff!
> 
> this might be a dumb question but why would we need `Color` in objc? wouldn't we only need that in swiftui?

@joannaquu ah I just reread what I wrote and it's a little unclear, sorry! I agree, Objective-C is fine only accessing UIKit. 

My intent was to point out that, unlike ObjC where everything works, in Swift we now have two methods with the same name that return different types. That's why I had to touch a couple files to add explicit variable types, like in `DrawerDemoController`:

```swift
// Error: Ambiguous use of 'color'
let backgroundColor = view.fluentTheme.color(.background3)

// Works
let backgroundColor: UIColor = view.fluentTheme.color(.background3)
```

I don't *love* this, but I don't like the idea of changing the UIKit signature either.

**Comment by joannaquu** [comment]
> > > > So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.
> > > 
> > > 
> > > That's correct. The Objective-C `-[MSFFluentTheme colorForToken:]` continues to be `UIColor` only. We could use different names for each, but I'm inclined towards leaving the existing name for `UIColor`, to minimize API breakage, and I couldn't think of a good name for the SwiftUI `Color` version. I'm open to suggestions though!
> > > I chose to make `DynamicColor` internal to keep the usage of our API simple: all our public APIs only deal with existing, understood color objects and structs. One less thing to have to learn to use our stuff!
> > 
> > 
> > this might be a dumb question but why would we need `Color` in objc? wouldn't we only need that in swiftui?
> 
> @joannaquu ah I just reread what I wrote and it's a little unclear, sorry! I agree, Objective-C is fine only accessing UIKit.
> 
> My intent was to point out that, unlike ObjC where everything works, in Swift we now have two methods with the same name that return different types. That's why I had to touch a couple files to add explicit variable types, like in `DrawerDemoController`:
> 
> ```swift
> // Error: Ambiguous use of 'color'
> let backgroundColor = view.fluentTheme.color(.background3)
> 
> // Works
> let backgroundColor: UIColor = view.fluentTheme.color(.background3)
> ```
> 
> I don't _love_ this, but I don't like the idea of changing the UIKit signature either.

ooh ok I see, one suggestion could be having `fluentTheme.color` and `fluentTheme.Color` (if that's even possible) but personally I don't mind having it return two different types. definitely seems cleaner than what we have now where we have to wrap it in `Color(uiColor:)` and also maybe needing explicit types will make client code cleaner

**Comment by mischreiber** [comment]
> > > > > So `fluentTheme.color(.token)` can now return either a `Color` or a `UIColor `? Also curious about why we opted to make `DynamicColor` internal.
> > > > 
> > > > 
> > > > That's correct. The Objective-C `-[MSFFluentTheme colorForToken:]` continues to be `UIColor` only. We could use different names for each, but I'm inclined towards leaving the existing name for `UIColor`, to minimize API breakage, and I couldn't think of a good name for the SwiftUI `Color` version. I'm open to suggestions though!
> > > > I chose to make `DynamicColor` internal to keep the usage of our API simple: all our public APIs only deal with existing, understood color objects and structs. One less thing to have to learn to use our stuff!
> > > 
> > > 
> > > this might be a dumb question but why would we need `Color` in objc? wouldn't we only need that in swiftui?
> > 
> > 
> > @joannaquu ah I just reread what I wrote and it's a little unclear, sorry! I agree, Objective-C is fine only accessing UIKit.
> > My intent was to point out that, unlike ObjC where everything works, in Swift we now have two methods with the same name that return different types. That's why I had to touch a couple files to add explicit variable types, like in `DrawerDemoController`:
> > ```swift
> > // Error: Ambiguous use of 'color'
> > let backgroundColor = view.fluentTheme.color(.background3)
> > 
> > // Works
> > let backgroundColor: UIColor = view.fluentTheme.color(.background3)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't _love_ this, but I don't like the idea of changing the UIKit signature either.
> 
> ooh ok I see, one suggestion could be having `fluentTheme.color` and `fluentTheme.Color` (if that's even possible) but personally I don't mind having it return two different types. definitely seems cleaner than what we have now where we have to wrap it in `Color(uiColor:)` and also maybe needing explicit types will make client code cleaner

Capitalizing the type would technically be a violation of the Microsoft Swift style guide, and would probably fail linting too.

For now I'm going to merge this as-is, but let's keep considering alternative names for these (up to and including renaming the existing API to be more explicit about `UIColor`) until we snap 0.28.

**Pull request merged**